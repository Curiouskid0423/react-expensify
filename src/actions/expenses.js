/**
 * Actions file for expenses.
 * @author Kevin Li
 */

import database from "../firebase/firebase";


/**
 * @desc A synchronous action generator that returns an object, based
 * on the params passed in by startAddExpense.
 * @param expense
 * @return {{type: string, expense: *}}
 */
export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
});

/**
 * @desc startAddExpense bootstraps the addExpense action dispatch.
 * @param description,  amount, createdAt, note
 * @return an async dispatch function call.
 */
export const startAddExpense = (
    { description = "", amount = 0, createdAt = 0, note = "" } = {}) => {
    return (dispatch, getState) => {
        const userId = getState().auth.uid;
        const expense = { description, amount, createdAt, note };
        // Returning this database call gives a Promise that may be tested on,
        // but the code can work completely fine even if the call isn't returned.
        return database.ref(`user/${userId}/expense`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    }
}

/**
 * Action for removing expense.
 * @param id
 * @returns {{id: *, type: string}}
 */
export const removeExpense = (id) => ({
        type: "REMOVE_EXPENSE", id
});

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const userId = getState().auth.uid;
        return database.ref(`user/${userId}/expense/${id}`).set(null).then(() => {
            dispatch(removeExpense(id));
        }, (error) => {
            console.log("Failed at removing an expense:", error);
        });
    }
}


/**
 * Action for editing expense based on provided info.
 * @param id
 * @param editObj
 * @returns {{edit: *, id: *, type: string}}
 */
export const editExpense = (id, editObj) => ({
        type: "EDIT_EXPENSE",
        id,
        edit: editObj
});

export const startEditExpense = (id, editObj = {}) => {
    return (dispatch, getState) => {
        const userId = getState().auth.uid;
         return database.ref(`user/${userId}/expense/${id}`)
             .update(editObj)
             .then(() => {
             dispatch(editExpense(id, editObj));
         });
    }
}

export const setExpenses = (expenses) => {
    return {
        type: "SET_EXPENSES",
        expenses
    };
}

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const userId = getState().auth.uid;
        return database.ref(`user/${userId}/expense`)
            .once("value")
            .then((snapshot) => {
                const lst = [];
                snapshot.forEach((child) => {
                    lst.push({
                        id: child.key,
                        ...child.val()
                    });
                });
                dispatch(setExpenses(lst));
            });
    }
}