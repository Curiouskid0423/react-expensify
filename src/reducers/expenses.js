/**
 * Expenses reducer file.
 * @author Kevin Li
 */

//TODO: Best practice to keep `reducer` just updating the value, and avoid computations.
//  Do the computations in action generators.
/**
 *  Expense reducer keeps track of the expenses array, i.e. only
 *  modifies the "expenses" portion of  the central state.
 * @param prevState
 * @param action Action to react to
 * @returns {*[]}
 */
const expenseReducer = (prevState = [], action) => {
    switch (action.type) {
        case "ADD_EXPENSE":
            return [
                ...prevState,
                action.expense
            ];
        case "REMOVE_EXPENSE":
            return prevState.filter((item) => item.id !== action.id);
        case "EDIT_EXPENSE":
            return prevState.map((item) => {
                if (item.id === action.id) {
                    item = {
                        ...item,
                        ...action.edit
                    }
                }
                return item;
            });
        default:
            return prevState;
    }
};

export default expenseReducer;