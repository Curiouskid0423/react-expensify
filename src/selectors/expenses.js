/**
 * Selector methods for expenses.
 * @author Kevin Li
 */

/**
 * Getter method for visible expenses based on filters.
 * @param expenses
 * @param filters
 * @returns {*}
 */
export const getVisibleExpenses = (expenses,
                            {text, sortBy, startDate, endDate} = filterDefault) => {
    const unSorted = expenses.filter((item) => {
        const textMatch = item.description.toLowerCase().includes(text.toLowerCase());
        const startMatch = typeof startDate !== "number" || startDate <= item.createdAt;
        const endMatch = typeof endDate !== "number" || endDate >= item.createdAt;
        return textMatch && startMatch && endMatch;
    });
    return sortExpenses(unSorted, sortBy);
}

/**
 * Helper method for sorting expense
 * @param target is the array to be sorted
 * @param sortBy is an indicator of what to sort by
 * @returns a sorted array of expenses
 */
export const sortExpenses = (target, sortBy) => {
    if (sortBy === "amount") {
        return target.sort((a, b) => b.amount - a.amount);
    }
    return target.sort((a, b) => b.createdAt - a.createdAt);
}