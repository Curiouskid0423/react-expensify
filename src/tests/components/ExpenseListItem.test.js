import React from "react";
import { shallow } from "enzyme";
import ExpenseItem from "../../components/ExpenseListItem";
import {expenses as defaultExp } from "../fixtures/expenses";

describe("ExpenseListItem Test Suite", () => {
   test("Test making an expense list with given object.", () => {
       const testExpense = defaultExp[1];
       const wrapper = shallow(<ExpenseItem
           key = {testExpense.id}
           {...testExpense}
       />);
       // First time matching snapshot is always going to succeed.
       // The idea here is to check the content whenever we have to.
       expect(wrapper).toMatchSnapshot();
   });
});