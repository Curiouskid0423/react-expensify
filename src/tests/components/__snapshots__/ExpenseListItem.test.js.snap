// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListItem Test Suite Should render an ExpenseListItem 1`] = `
<div>
  <Link
    to="/edit/2"
  >
    <h4>
      Louisa Coffee
    </h4>
  </Link>
  <h4>
    $7,000.00
     -- 
    Jan 3rd 1970
  </h4>
</div>
`;
